src/databases.py

```diff
@@ -6,37 +6,67 @@
 import logging
 import typing
 
-from charms.data_platform_libs.v0.data_interfaces import DatabaseRequires
+import ops.framework
+import yaml
+from charms.data_platform_libs.v0.data_interfaces import DatabaseRequires, DatabaseRequiresEvent
+from ops.charm import CharmBase
 
-from constants import FLASK_DATABASE_NAME, FLASK_SUPPORTED_DATABASES
-from exceptions import InvalidDatabaseRelationDataError
-
-if typing.TYPE_CHECKING:
-    from charm import FlaskCharm
+from constants import FLASK_CONTAINER_NAME, FLASK_DATABASE_NAME, FLASK_SERVICE_NAME
+from exceptions import InvalidDatabaseRelationDataError, PebbleNotReadyError
 
 logger = logging.getLogger(__name__)
 
 
-class Databases:  # pylint: disable=too-few-public-methods
+# We need to derive from ops.framework.Object to subscribe to callbacks
+# from ops.framework. See: https://github.com/canonical/operator/blob/main/ops/framework.py#L782
+class Databases(ops.framework.Object):  # pylint: disable=too-few-public-methods
     """A class handling databases relations and state.
 
     Attrs:
         _charm: The main charm. Used for events callbacks
         _databases: A dict of DatabaseRequires to store relations
     """
 
-    def __init__(self, charm: "FlaskCharm"):
+    def __init__(self, charm: CharmBase):
         """Initialize a new instance of the Databases class.
 
         Args:
             charm: The main charm. Used for events callbacks
         """
+        # The following is necessary to be able to subscribe to callbacks from ops.framework
+        super().__init__(charm, "databases")
         self._charm = charm
+
+        with open("metadata.yaml", encoding="utf-8") as metadata_fo:
```

Comment:
pathlib provides an API `.read_text` which makes this much easier: https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text

Reply:
Changed